examples/auto-gen-ca-and-server-tls.rs

```diff
@@ -0,0 +1,217 @@
+//! Auto-generate a CA certificate and a server certificate signed
+//! by that CA, and start a TLS server and client using them.
+//!
+//! Run with:
+//!
+//!     $ cargo run --example auto-gen-ca-and-server-tls --features=x509-parser
+//!
+//! This doesn't start a network service (not even on localhost).
+//! Instead, it creates an in-memory TLS server and an in-memory
+//! TLS client in two separate threads in the same process.
+
+use std::error::Error;
+use std::io::{BufRead, BufReader, Write};
+use std::sync::Arc;
+use std::thread;
+use rustls;
+use rustls::{PrivateKey, RootCertStore, ServerName};
+use rustls::server::Acceptor;
+
+use rcgen::{BasicConstraints, CertificateSigningRequest, DnType, IsCa, SanType, KeyUsagePurpose, DistinguishedName, CertificateParams, Certificate, RcgenError};
+
+const SAN: &str = "example-server";
+
+fn read_line<A: BufRead>(reader: &mut A) -> Result<String, std::io::Error> {
+	let mut buffer = String::new();
+
+	reader.read_line(&mut buffer)?;
+
+	Ok(buffer.trim_end().into())
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let ca = gen_cert_for_ca()?;
+	let rustls_ca = rustls::Certificate(ca.serialize_der().unwrap());
+
+    println!("CA private key:\n{}", ca.serialize_private_key_pem());
+    println!();
+    println!("CA certificate:\n{}", ca.serialize_pem()?);
+
+    let host_cert = gen_cert_for_server(&ca)?;
+	let host_cert_rustls = rustls::Certificate(host_cert.signed_certificate_der);
+	let host_pk = PrivateKey(host_cert.private_key_der);
+
+    println!("Preparing to verify with tls");
+
+    let leaf_then_ca = vec![host_cert_rustls, rustls_ca.clone()];
+
+    println!();
+
+	// Does not work with either buffered or unbuffered bipipe, even when we use the same message
+	// pattern as with the nativetls example, where client sends then reads, and server reads and
+	// then sends. Client blocks sending a message; server blocks reading a message.
+    let (mut p_client, mut p_server) = pipe::bipipe_buffered();
```

Comment:
Note that there's really no need to use a pipe thing here. You can just directly use a `Vec<u8>` buffer (or two) that one side writes into and the other side then reads from.

Reply:
The vectors couldn't be moved into the server thread and client threads, because it would require moving each vector twice.